File: scripts/find_version_commit.py
#!/usr/bin/python3

# @author Bodo (Hugo) Barwich
# @version 2023-02-21
# @package TextSanitizer
# @subpackage scripts/cargo_version.py

# This Module parses the Git History to find the Merge Commit for a given Commit Hash
#

import sys
import os
import os.path
import toml
import json
from git import Repo, GitCommandError


# ==============================================================================
# Auxiliary Functions

def git_history(module_name, git):
    history_result = {'success': True, 'commits': []}
    history_info = None
    history_lines = []

    try:
        history_info = git.log('-50')
    except GitCommandError as e:
        if not module_quiet:
            print(
                "script '{}' - Cargo File '{}': Blame File failed!".format(
                    module_name, filepath), file=sys.stderr)
            print("script '{}' - Cargo File Exception Message: {}".format(
                module_name, str(e)), file=sys.stderr)

        history_result['success'] = False

    if module_debug:
        print("history info:'{}'".format(history_info))

    if history_info is not None:
        history_lines = history_info.split("\n")

    commit_idx = 0

    for line in history_lines:
        if line[0: 6] == 'commit':
            commit_hash = line.split(' ')[1]
            commit = {'hash': commit_hash,
                      'hash_short': commit_hash[0: 7],
                      'raw': [],
                      'index': commit_idx}
            commit['raw'].append(line)

            if module_debug:
                print("commit dict:'{}'".format(str(commit)))

            history_result['commits'].append(commit)

            commit_idx += 1

        else:
            commit['raw'].append(line)
            colon_pos = line.find(':')

            if colon_pos != -1:
                line_key = line[0: colon_pos].lower()
                line_value = line[colon_pos + 1: len(line)].lstrip()
                author_info = None
                merge_info = None

                if line_key == 'author':
                    author_info = parse_commit_author(line_value)
                elif line_key == 'merge':
                    merge_commits = line_value.split(' ')
                    merge_info = {
                        'dest': merge_commits[0],
                        'origin': merge_commits[1]}

                if author_info is not None:
                    commit[line_key] = author_info
                elif merge_info is not None:
                    commit[line_key] = merge_info
                else:
                    commit[line_key] = line_value

    return history_result


def parse_commit_author(author_value):
    author = None
    email_pos = author_value.find('<')

    if email_pos != -1:
        author = {'name': author_value[0: email_pos - 1].rstrip(),
                  'email': author_value[email_pos + 1: len(author_value)].rstrip('>')}

    return author


def find_merge_commit(history, commit):
    commit_idx = len(history) - 1
    commit_search = None
    commit_merge = None

    if 'index' in commit:
        commit_idx = commit['index']

    while commit_idx >= 0 and commit_merge is None:
        commit_search = history[commit_idx]

        if 'merge' in commit_search:
            commit_merge = commit_search
        else:
            commit_idx -= 1

    return commit_merge


# ==============================================================================
# Executing Section


# ------------------------
# Script Environment

module_file = ''
module_path = os.path.abspath(__file__)
main_dir = ''
work_dir = ''


slash_pos = module_path.rfind('/', 0)

if slash_pos != -1:
    work_dir = module_path[0: slash_pos + 1]
    module_file = module_path[slash_pos + 1: len(module_path)]
else:
    module_file = module_path

if work_dir != '':
    slash_pos = work_dir.rfind('/', 0, -1)
    if slash_pos != -1:
        main_dir = work_dir[0: slash_pos + 1]
    else:
        main_dir = work_dir


# ------------------------
# Script Parameter

commit_search = []
module_output = 'plain'
module_debug = False
module_quiet = False
module_res = 0

for arg in sys.argv:
    if arg[0: 2] == '--':
        arg = arg[2: len(arg)]
        if arg in ['plain', 'json']:
            module_output = arg
        elif arg == 'debug':
            module_debug = True
        elif arg == 'quiet':
            module_quiet = True

    elif arg[0] == '-':
        arg = arg[1: len(arg)]
        for idx in range(0, len(arg)):
            if arg[idx] == 'd':
                module_debug = True
            elif arg[idx] == 'q':
                module_quiet = True
    else:
        if arg.rfind(module_file, 0) == -1:
            if arg[0] == '^':
                arg = arg[1: len(arg)]

            if len(arg) > 7:
                commit_search.append(arg[0: 7])
            else:
                commit_search.append(arg)

if module_debug:
    print(
        "script '{}' - Commit Searches:\n{}".format(module_file, str(commit_search)))
    print(
        "script '{}' - Search Output: '{}'".format(module_file, module_output))

if len(commit_search) == 0:
    print(
        "script '{}' - Commit Hash is missing!".format(module_file))

    module_res = 3

# ------------------------
# Parse the Git History


repo = Repo('.git')
git = repo.git

history_result = git_history(module_file, git)
history_commits = {}
commits_res = {}

if module_debug:
    print(
        "script '{}' - History Result:\n{}".format(module_file, str(history_result)))

if not history_result['success']:
    if not module_quiet:
        print("script '{}' - Git History Parsing: Parsing History has failed!".format(
            module_file), file=sys.stderr)

    module_res = 1

for commit in history_result['commits']:
    if 'hash_short' in commit:
        history_commits[commit['hash_short']] = commit

for search in commit_search:
    commits_res[search] = 0

    if search in history_commits:
        commits_res[search] = {
            'origin': history_commits[search],
            'merge': find_merge_commit(
                history_result['commits'],
                history_commits[search])}

# ------------------------
# Print the Commit Result

if module_output == 'plain':
    if len(commits_res) > 0:
        print("script '{}' - Version Commits:".format(module_file))

        for search in commits_res:
            if commits_res[search] != 0:
                if commits_res[search]['merge'] is not None:
                    print(
                        "{}/{} by '{}/{}'".format(
                            commits_res[search]['merge']['hash_short'],
                            commits_res[search]['merge']['hash'],
                            commits_res[search]['merge']['author']['name'],
                            commits_res[search]['merge']['author']['email']))
                else:
                    print("{}/{} by '{}/{}'".format(search,
                                                    commits_res[search]['origin']['hash'],
                                                    commits_res[search]['origin']['author']['name'],
                                                    commits_res[search]['origin']['author']['email']))

            else:
                print("{} - no entry found".format(search))

                module_res = 1

elif module_output == 'json':
    print("{}".format(json.dumps(commits_res)))

else:
    print(
        "script '{}' - Version Commits:\n{}".format(module_file, str(commits_res)))


if module_debug:
    print("script '{}': Script finished with [{}]".format(
        module_file, module_res))


sys.exit(module_res)

File: scripts/find_version_commit.pl
#!/usr/bin/perl

# @author Bodo (Hugo) Barwich
# @version 2023-07-05
# @package Process::SubProcess
# @subpackage scripts/find_version_commit.pl

# This Module parses the Git History to find the Merge Commit for the Version
#

use strict;
use warnings;

use Getopt::Long;
use Path::Tiny qw(path);
use JSON qw(encode_json);
use Data::Dump qw(dump);
use Git;

# ==============================================================================
# Auxiliary Functions

sub git_history {
    my ( $module_name, $maindir ) = @_;
    my %history_result = ( 'success' => 1, 'commits' => [] );

    my @log_lines  = ();
    my $commit     = undef;
    my $commit_idx = 0;

    eval {
        my $repo = Git->repository( Repository => $maindir . '/.git' );

        @log_lines = $repo->command( 'log', '-50' );
    };

    if ($@) {
        print STDERR
          "script '$module_name' - Git History: Git Log Command failed!\n";
        print STDERR
          "script '$module_name' - Git Log Exception Message: '$@'\n";

        @log_lines = ();

        $history_result{'success'} = 0;
    }

    $history_result{'success'} = 0 if ( scalar @log_lines == 0 );

    foreach my $line (@log_lines) {
        if ( $line =~ qr/^commit (.*)$/i ) {

            $commit = {
                'hash'       => $1,
                'hash_short' => substr( $1, 0, 7 ),
                'raw'        => $line . "\n",
                'index'      => $commit_idx
            };

            push @{ $history_result{'commits'} }, ($commit);

            $commit_idx++;
        }
        else {
            $commit->{'raw'} .= $line . "\n";

            if ( index( $line, ':' ) != -1 ) {
                if ( $line =~ qr/^([^:]+): (.*)/ ) {
                    $commit->{ lc $1 } = $2;
                }
            }
        }
    }

    foreach $commit ( @{ $history_result{'commits'} } ) {
        if ( defined $commit->{'author'} ) {
            if ( $commit->{'author'} =~ qr/^([^<]+) <([^>]+)>/ ) {
                $commit->{'author'} = { 'name' => $1, 'email' => $2 };
            }
        }

        if ( defined $commit->{'date'} ) {
            $commit->{'date'} =~ s/^\s+//;
        }
    }

    return \%history_result;
}

sub find_merge_commit {
    my ( $history, $commit ) = @_;

    my $commit_idx    = scalar($history) - 1;
    my $commit_search = undef;
    my $commit_merge  = undef;

    $commit_idx = $commit->{'index'} if ( defined $commit->{'index'} );

    while ( $commit_idx >= 0 && !defined $commit_merge ) {
        $commit_search = $history->[$commit_idx];

        if ( defined $commit_search->{'merge'} ) {
            $commit_merge = $commit_search;
        }
        else {
            $commit_idx--;
        }
    }

    return $commit_merge;
}

# ==============================================================================
# Executing Section

# ------------------------
# Script Environment

my $module_file = path($0)->basename;
my $path        = path($0)->parent->absolute;
my $maindir     = $path->parent;

# ------------------------
# Script Parameter

my @rqcommits     = ();
my $output_format = 'plain';
my $debug         = 0;
my $quiet         = 0;

my $ierr = 0;

GetOptions(
    'f|format=s' => \$output_format,
    'd|debug'    => \$debug,
    'q|quiet'    => \$quiet
);
@rqcommits = @ARGV;

@rqcommits =
  map { index( $_, '^' ) == 0 ? substr( $_, 1, length($_) ) : $_ } @rqcommits;
@rqcommits = map { length($_) > 7 ? substr( $_, 0, 7 ) : $_ } @rqcommits;

my %commits_res = ();

my $history_result = git_history( $module_file, $maindir->stringify );

my %history_commits = map { $_->{'hash_short'} => $_ }
  grep { defined $_->{'hash_short'} } @{ $history_result->{'commits'} };

if ($debug) {
    print "hist res dmp:\n", dump($history_result), "\n";
}

my $search = '';

foreach $search (@rqcommits) {
    $commits_res{$search} = 0;

    if ( defined $history_commits{$search} ) {
        $commits_res{$search} = {
            'origin' => $history_commits{$search},
            'merge'  => find_merge_commit(
                $history_result->{'commits'},
                $history_commits{$search}
            )
        };
    }
}

# ------------------------
# Print the Commit Result

if ( $output_format eq 'plain' ) {
    if ( scalar keys %commits_res > 0 ) {
        print "script '$module_file' - Version Commits:\n";

        foreach $search ( keys %commits_res ) {
            if ( ref $commits_res{$search} ne '' ) {
                if ( defined $commits_res{$search}->{'merge'} ) {
                    printf(
                        "%s/%s by '%s/%s'\n",
                        $commits_res{$search}->{'merge'}->{'hash_short'},
                        $commits_res{$search}->{'merge'}->{'hash'},
                        $commits_res{$search}->{'merge'}->{'author'}->{'name'},
                        $commits_res{$search}->{'merge'}->{'author'}->{'email'}
                    );
                }
                else {
                    printf(
                        "%s/%s by '%s/%s'\n",
                        $search,
                        $commits_res{$search}->{'origin'}->{'hash'},
                        $commits_res{$search}->{'origin'}->{'author'}->{'name'},
                        $commits_res{$search}->{'origin'}->{'author'}->{'email'}
                    );
                }
            }
            else {
                print $search, " - no entry found\n";

                $ierr = 1;
            }
        }
    }

}
elsif ( $output_format eq 'json' ) {
    print encode_json( \%commits_res );
}
else {
    print "script '$module_file' - Version Commits:\n", dump( \%commits_res ),
      "\n";
}

if ($debug) {
    print "script '$module_file': Script finished with [$ierr]\n";
}

exit $ierr;
